lib/rust/flexer/tests/flexer_lexer_definition_test.rs

```diff
@@ -0,0 +1,229 @@
+//! This file contains tests for the intended defnition code of a lexer using the flexer, based on
+//! the following small language.
+//!
+//! The language here is being defined as follows:
+//!
+//! a-word      = 'a'+;
+//! b-word      = 'b'+;
+//! word        = a-word | b-word;
+//! space       = ' ';
+//! spaced-word = space, word;
+//! language    = word, spaced-word*;
+
+use flexer::prelude::*;
+use flexer::automata::dfa::DFA;
+use flexer::automata::nfa::NFA;
+use flexer::group::Group;
+use lazy_reader::{BookmarkId, LazyReader, Reader};
+use flexer::{FlexerState, Flexer};
+use lazy_reader::decoder::DecoderUTF8;
+use flexer::automata::pattern::Pattern;
+
+
+// ===========
+// === AST ===
+// ===========
+
+/// A very simple AST, sufficient for the simple lexer being defined.
+#[derive(Clone,Debug,PartialEq)]
+pub enum AST {
+    /// A word from the input, consisting of a sequence of all `a` or all `b`.
+    Word(String),
+    /// A token that the lexer is unable to recognise.
+    Unrecognised(String)
+}
+
+
+
+// =============
+// === Lexer ===
+// =============
+
+#[allow(missing_docs)]
+#[derive(Clone,Debug)]
+pub struct Lexer {
+    groups: Vec<Group>
+}
+
+#[allow(missing_docs)]
+impl Lexer {
+    pub fn new() -> Self {
+        let groups = Vec::new();
+        Lexer{groups}
+    }
+
+    // TODO [AA] Parent groups.
+    pub fn define_group(&mut self,name:&str) -> &mut Group {
+        let id    = self.groups.len();
+        let group = Group::new(id,String::from(name),None);
+        self.groups.push(group);
+        self.groups.get_mut(id).expect("Has just been pushed so should always exist.")
+    }
+
+    pub fn specialize(&self) -> String {
+        let group_nfa:Vec<NFA>  = self.groups.iter().map(|group|group.into()).collect();
+        let _group_dfa:Vec<DFA> = group_nfa.iter().map(|nfa|nfa.into()).collect();
+        let str = String::new();
+        str
+    }
+}
+
+#[derive(Debug)]
+#[allow(missing_docs)]
+pub struct TestLexer<Reader:LazyReader> {
+    lexer: Flexer<TestState,AST,Reader>
+}
+
+/// The definition?
+#[allow(missing_docs)]
+impl<Reader:LazyReader> TestLexer<Reader> {
+    pub fn new(reader:Reader) -> Self {
+        let lexer = Flexer::new(reader);
+        TestLexer{lexer}
+    }
+}
+
+/// Implementation helpers.
+#[allow(missing_docs)]
+impl<Reader:LazyReader> TestLexer<Reader> {
+    pub fn def_on_first_word_str(&mut self) {
+        let str = self.current_match.clone();
+        let ast = AST::Word(str);
```

Comment:
I kniw this is ock, but in real code, constructors should be responsible for cloning things (by accepting `impl Into<String>`.

Reply:
How do you do this? As far as I understand it, an enum variant (which is what `AST::Word` is) can't accept `impl Trait` as is type.

Reply:
Note to self.

```rust
enum AST {
    Unrecognized {name:String},
    Word {name:String}
}

fn Unrecognized(name:impl Into<String>) -> AST { ... }
```